### PR TITLE
#12087 - Fix Bulk Edit update when M2M operations are present.

### LINKED ISSUE #12087: Interface VLANs Not Updating Properly
### NetBox version

v3.4.7

### Python version

3.10

### Steps to Reproduce

1. Select multiple device interfaces (that already have mode configured as tagged, an assigned untagged VLAN, and exactly one assigned tagged VLAN) and initiate a bulk edit
2. Change the untagged VLAN
3. Change the tagged VLAN
4. Apply

### Expected Behavior

Both the untagged and tagged VLAN of the interfaces are updated.

### Observed Behavior

The interfaces are only updated with the supplied tagged VLAN; the untagged VLAN remains unchanged.

Before:
image

Attempted Changes:
image

After:
image

### PR DESCRIPTION
**Pull request by DanSheps**
### Fixes: #12087

* Alters Bulk Update logic to perform M2M operations after the save

### CODE REVIEW

File: netbox/netbox/views/generic/bulk_views.py [hunk 1]
```diff
@@ -511,25 +526,11 @@ def _update_objects(self, form, request):
...
-                    # This form field is used to modify a field rather than set its value directly
-                    model_field = None
-
+            for name, model_field in model_fields.items():
                 # Handle nullification
                 if name in form.nullable_fields and name in nullified_fields:
-                    if isinstance(model_field, ManyToManyField):
-                        getattr(obj, name).set([])
-                    else:
+                    if not isinstance(model_field, ManyToManyField):
```
Comment (DanSheps) [inline netbox/netbox/views/generic/bulk_views.py:532]: Might be able to just dump this as model_fields shouldn't contain m2m fields anymore.

File: netbox/netbox/views/generic/bulk_views.py [hunk 2]
```diff
@@ -503,6 +503,21 @@ def _update_objects(self, form, request):
...
         nullified_fields = request.POST.getlist('_nullify')
         updated_objects = []
+        model_fields = {}
+        m2m_fields = {}
+
+        # Update standard fields. If a field is listed in _nullify, delete its value.
+        for name in standard_fields:
+            try:
+                model_field = self.queryset.model._meta.get_field(name)
+                if isinstance(model_field, ManyToManyField):
```
Comment (DanSheps) [inline netbox/netbox/views/generic/bulk_views.py:513]: Should maybe change this to `(ManyToManyField,ManyToManyRel)`

File: netbox/netbox/views/generic/bulk_views.py [hunk 3]
```diff
@@ -503,6 +503,21 @@ def _update_objects(self, form, request):
...
+                    m2m_fields[name] = model_field
+                else:
+                    model_fields[name] = model_field
+
+            except FieldDoesNotExist:
+                # This form field is used to modify a field rather than set its value directly
+                model_field = None
```
Comment (DanSheps) [inline netbox/netbox/views/generic/bulk_views.py:520]: Should do `model_fields[name] = None`

File: netbox/netbox/views/generic/bulk_views.py [hunk 4]
```diff
@@ -503,6 +503,21 @@ def _update_objects(self, form, request):
         ]
         nullified_fields = request.POST.getlist('_nullify')
         updated_objects = []
+        model_fields = {}
+        m2m_fields = {}
+
+        # Update standard fields. If a field is listed in _nullify, delete its value.
```
Comment (jeremystretch) [inline netbox/netbox/views/generic/bulk_views.py:509]: Don't forget to update the comments appropriately. (This has been duplicated from below.)

File: netbox/netbox/views/generic/bulk_views.py [hunk 5]
```diff
@@ -547,6 +547,13 @@ def _update_objects(self, form, request):
             obj.save()
             updated_objects.append(obj)
 
+            # Handle M2M fields after save
+            for name, m2m_field in m2m_fields.items():
+                if name in form.nullable_fields and name in nullified_fields:
+                    getattr(obj, name).set([])
```
Comment (jeremystretch) [inline netbox/netbox/views/generic/bulk_views.py:553]:
```suggestion
                    getattr(obj, name).clear()
```

**Review by jeremystretch — CHANGES_REQUESTED** [review]
Thanks for taking this @DanSheps!

File: netbox/netbox/views/generic/bulk_views.py (hunk 4 above)
Comment (DanSheps) [reply netbox/netbox/views/generic/bulk_views.py]: Fixed up that comment, whoops

**Pull request merged**